fastavro/_write_py.py

```diff
@@ -296,89 +312,174 @@ def write_map(fo, datum, schema):
+    return (
             isinstance(datum, Mapping) and
             all(is_str(k) for k in iterkeys(datum)) and
-            all(validate(v, schema['values']) for v in itervalues(datum))
-        )
+            all(validate(datum=v, schema=schema['values'],
+                         field=schema.get('name'),
+                         raise_errors=raise_errors) for v in itervalues(datum))
+    )
+
 
-    if record_type in ('record', 'error', 'request',):
-        return (
+def validate_record(datum, schema, raise_errors=False):
+    return (
             isinstance(datum, Mapping) and
             all(
-                validate(datum.get(f['name'], f.get('default')), f['type'])
+                validate(datum=datum.get(f['name'], f.get('default')),
+                         schema=f['type'],
+                         field=schema.get('name'),
+                         raise_errors=raise_errors)
                 for f in schema['fields']
             )
-        )
+    )
 
-    if record_type in SCHEMA_DEFS:
-        return validate(datum, SCHEMA_DEFS[record_type])
 
-    raise ValueError('unkown record type - %s' % record_type)
+def validate_union(datum, schema, raise_errors=False):
+    if isinstance(datum, tuple):
+        (name, datum) = datum
+        for candidate in schema:
+            if extract_record_type(candidate) == 'record':
+                if name == candidate["name"]:
+                    return validate(datum, schema=candidate,
+                                    field=None,
+                                    raise_errors=raise_errors)
+        else:
+            return False
+
+    errors = []
+    for s in schema:
+        try:
+            ret = validate(datum, schema=s,
+                           field=None,
+                           raise_errors=raise_errors)
+            if ret:
+                # We exit on the first passing type in Unions
+                return True
+        except ValidationException as e:
+            errors.append(e.message)
+    if raise_errors:
+        raise ValidationErrors(errors)
+    return False
+
+
+VALIDATORS = {
+    'null': validate_null,
+    'boolean': validate_boolean,
+    'string': validate_string,
+    'int': validate_int,
+    'long': validate_long,
+    'float': validate_float,
+    'double': validate_float,
+    'bytes': validate_bytes,
+    'fixed': validate_fixed,
+    'enum': validate_enum,
+    'array': validate_array,
+    'map': validate_map,
+    'union': validate_union,
+    'error_union': validate_union,
+    'record': validate_record,
+    'error': validate_record,
+    'request': validate_record
+}
+
+
+def validate(datum, schema, field=None, raise_errors=False):
+    """Determine if a python datum is an instance of a schema."""
+    record_type = extract_record_type(schema)
+    result = None
+
+    if hasattr(schema, 'get'):
+        if field is not None:
+            ns_field = '.'.join([field, schema.get('name')])
```

Comment:
Either I'm not understanding the intention of this correctly, or I don't think it's working correctly. I changed your `test_validate_int_in_string_raises` to actually raise the exception like so:

```python
    # with pytest.raises((ValidationException,)):
    #     validation_raise(schema, *records)
    validation_raise(schema, *records)
```

And the result is `fastavro._write_common.ValidationException: missingerror is 11 of type <class 'int'> expected string`

If I understand correctly, I think the intention is that the message should say `missingerror.str`, not just `missingerror`

Reply:
Ah because this is only joining the `namespace.name`, not the `namespace.name.type`, should be a simple fix.